src/promise.test.ts

```diff
@@ -0,0 +1,49 @@
+import { createDeferredPromise } from './promise';
+
+describe('Promise utilities', () => {
+  describe('createDeferredPromise', () => {
+    it('creates a deferred promise that resolves when resolve is called', async () => {
+      const { promise, resolve } = createDeferredPromise();
+
+      resolve();
+
+      expect(await promise).toBeUndefined();
+    });
+
+    it('creates a deferred promise that rejects when reject is called', async () => {
+      const { promise, reject } = createDeferredPromise();
+      const mockError = new Error('test error');
+
+      reject(mockError);
+
+      await expect(promise).rejects.toThrow('test error');
+    });
+
+    it('ignores subsequent calls to reject or resolve', async () => {
+      const { promise, reject, resolve } = createDeferredPromise();
+      resolve();
+      await promise;
+
+      expect(() => reject(new Error('test error'))).not.toThrow();
+      expect(() => resolve()).not.toThrow();
+    });
+
+    describe('when suppressUnhandledRejection is set', () => {
```

Comment:
I tried writing a test for when this was false, but there was no way to catch the `unhandledRejection` event. Apparently in the environment Jest runs tests in, the event always bubbles up to the parent process. Attaching an event handler for that event has no effect here. And there was no easy way to attach an event handler to the parent process, I think it'd require patching Jest or creating a custom test environment.